control/runner.go

```diff
@@ -306,9 +306,87 @@ func (r *runner) HandleGomitEvent(e gomit.Event) {
 		if err != nil {
 			return
 		}
+	case *control_event.UnloadPluginEvent:
+		// On plugin unload,  find the key and pool info for the plugin being unloaded.
+		r.availablePlugins.Lock()
+		var pool strategy.Pool
+		var k string
+		for key, p := range r.availablePlugins.table {
+			tnv := strings.Split(key, ":")
+			if core.PluginType(v.Type).String() == tnv[0] && v.Name == tnv[1] && v.Version == p.Version() {
+				pool = p
+				k = key
+				break
+			}
+		}
+
+		r.availablePlugins.Unlock()
+		if pool == nil {
+			return
+		}
+		// Check for the highest lower version plugin and move subscriptions that
+		// are not bound to a plugin version to this pool.
+		plugin, err := r.pluginManager.get(fmt.Sprintf("%s:%s:%d", core.PluginType(v.Type).String(), v.Name, -1))
+		if err != nil {
+			return
+		}
+		newPool, err := r.availablePlugins.getOrCreatePool(plugin.Key())
+		if err != nil {
+			return
+		}
+		subs := pool.MoveSubscriptions(newPool)
+		// Start new plugins in newPool if needed
+		if newPool.Eligible() {
```

Comment:
NewPool should be eligible under most scenarios I can think of. If it wasn't that would mean that at some point between starting our unload event handling and this line the pool for that plugin was initialized. I think the way this might happen is if a task was submitted that had subscriptions to this specific pool and that was processed between unloading the plugin and this line.

Deleting the plugin from available plugins still needs to happen regardless I think. We need to remove an unloaded plugin from the list of available plugins no matter what happens with moving the subscriptions since it should not be available (since we were told to unload it).